packages/loopring_v3/contracts/amm/libamm/AmmBlockReceiver.sol

```diff
@@ -4,124 +4,135 @@ pragma solidity ^0.7.0;
 pragma experimental ABIEncoderV2;
 
 import "../../core/impl/libtransactions/BlockReader.sol";
+import "../../lib/MathUint.sol";
 import "./AmmData.sol";
+import "./AmmExchange.sol";
 import "./AmmExitProcess.sol";
 import "./AmmJoinProcess.sol";
+import "./AmmPoolToken.sol";
 import "./AmmUpdateProcess.sol";
 
 
 /// @title AmmBlockReceiver
 library AmmBlockReceiver
 {
+    using AmmExchange       for AmmData.State;
     using AmmExitProcess    for AmmData.State;
     using AmmJoinProcess    for AmmData.State;
+    using AmmPoolToken      for AmmData.State;
     using AmmUpdateProcess  for AmmData.State;
     using BlockReader       for ExchangeData.Block;
+    using MathUint          for uint;
 
-    function beforeBlockSubmitted(
+    function beforeBlockSubmission(
         AmmData.State      storage S,
         ExchangeData.Block memory  _block,
-        uint                       txIdx,
-        bytes              memory  auxiliaryData
+        bytes              memory  poolTxData,
+        uint                       txIdx
         )
         public
         returns (uint)
     {
-        AmmData.PoolTransaction[] memory poolTransactions = abi.decode(
-            auxiliaryData,
-            (AmmData.PoolTransaction[])
-        );
-
-        // Cache the domain seperator to save on SLOADs each time it is accessed.
-        uint size = S.tokens.length;
-        AmmData.Context memory ctx = AmmData.Context({
-            _block: _block,
-            exchange: S.exchange,
-            exchangeDepositContract: address(S.exchange.getDepositContract()),
-            txIdx: txIdx,
-            domainSeparator: S.domainSeparator,
-            exchangeDomainSeparator: S.exchange.getDomainSeparator(),
-            ammActualL2Balances: new uint96[](size),
-            ammExpectedL2Balances: new uint96[](size),
-            numTransactionsConsumed: 0,
-            tokens: S.tokens,
-            size: size,
-            poolTokenBase: AmmData.LP_TOKEN_BASE(),
-            poolTokenInitialSupply: AmmData.LP_TOKEN_INITIAL_SUPPLY()
-        });
+        AmmData.Context memory ctx = getContext(S, _block, txIdx);
+
+        // Question(brecht): is it OK not to check the following for every poolTx?
```

Comment:
In theory the exchange owner can be the owner of multiple exchanges, and we have to make sure the block is for the correct exchange. But yeah, seems a bit far fetched, but not impossible. :)

Reading the header is pretty cheap so I don't think there's a problem keeping it.